modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,113 @@
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.either._
+import cats.syntax.eq._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.semigroupal._
+import cats.syntax.traverse._
+import cats.syntax.unorderedFoldable._
+
+import org.tessellation.dag.l0.infrastructure.snapshot.{DAGEvent, StateChannelEvent}
+import org.tessellation.ext.kryo._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.node.shared.domain.event.EventCutter
+import org.tessellation.schema.Block
+import org.tessellation.security.hash._
+import org.tessellation.security.{Hashed, Hasher}
+
+import eu.timepit.refined.types.numeric.PosInt
+
+object GlobalSnapshotEventCutter {
+  def make[F[_]: Async: KryoSerializer: Hasher](maxBinarySizeBytes: PosInt): EventCutter[F, StateChannelEvent, DAGEvent] =
+    new EventCutter[F, StateChannelEvent, DAGEvent] {
+
+      def cut(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): F[(List[StateChannelEvent], List[DAGEvent])] =
+        doesNotExceedMaxSize(scEvents, dagEvents)
+          .ifM((scEvents, dagEvents).pure[F], removeEvents(scEvents, dagEvents))
+
+      def doesNotExceedMaxSize(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): F[Boolean] =
+        (scEvents.toBinary.liftTo[F], dagEvents.toBinary.liftTo[F])
+          .mapN((a, b) => a.length + b.length <= maxBinarySizeBytes.value)
+
+      def removeEvents(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): F[(List[StateChannelEvent], List[DAGEvent])] =
+        (
+          scEvents.traverse(evt => evt.pure[F].product(StateChannelEventWithFee(evt))).map(_.toMap),
+          dagEvents.traverse(evt => evt.pure[F].product(DAGBlockEventWithFee(evt))).map(_.toMap)
+        ).flatMapN {
+          case (scEventsMap, dagEventsMap) =>
+            def removeOne(scEvents: List[StateChannelEvent], dagEvents: List[DAGEvent]): (List[StateChannelEvent], List[DAGEvent]) =
+              (scEvents, dagEvents) match {
+                case (Nil, Nil)    => (Nil, Nil)
+                case (_ :: t, Nil) => (t, Nil)
+                case (Nil, _ :: t) => (Nil, t)
+                case (h1 :: t1, h2 :: t2) =>
+                  Ordering[EventWithFee].compare(scEventsMap(h1), dagEventsMap(h2)) match {
+                    case n if (n == -1) || (n == 0 && t1.length >= t2.length) => (t1, dagEvents)
+                    case _                                                    => (scEvents, t2)
+                  }
+              }
+
+            val scEventsSorted = scEvents.sortWith((left, right) => Ordering[EventWithFee].lt(scEventsMap(left), scEventsMap(right)))
+            val dagEventsSorted = dagEvents.sortWith((left, right) => Ordering[EventWithFee].lt(dagEventsMap(left), dagEventsMap(right)))
+
+            (scEventsSorted, dagEventsSorted).tailRecM {
+              case (Nil, Nil) => MonadThrow[F].raiseError(new IllegalStateException("No events to cut"))
```

Comment:
Maybe it should just return empty lists? You could also log it, but maybe throwing is not neccessary? Although I get it that you don't expect it to ever fail with a correct config, but take into account that someone can use really low numbers and no events would pass the limit - in this situation empty lists are correct results right?

Reply:
Is it really valid to remove events to the point they're all gone? I thought the objective was to reduce the no. of events when they're unreasonably large, so if the max size is small then the configuration is wrong.

If you think that's better I can have it return empty list, but if we do I think we should have a larger minimum for the size than 0.

Reply:
I think we can have this function to be a more generic thing. If someone really sets the limit too low for their use case than that's their responsibility. Anyway I wouldn't throw if the result is empty lists - the function did what it was told to do, so the error is not it's fault. If we need some validation on the limit then I would impose it with a custom refined type used in the config.

Reply:
And I would still leave the cutter parameter `maxBinarySizeBytes ` to be a `PosInt`.

Reply:
@TheMMaciek Changed to not throw.